src/algebra/hom/units.lean

```diff
@@ -103,31 +108,166 @@ def to_hom_units {G M : Type*} [group G] [monoid M] (f : G →* M) : G →* Mˣ
 
-@[to_additive] lemma is_unit.map {F : Type*} [monoid M] [monoid N] [monoid_hom_class F M N]
-  (f : F) {x : M} (h : is_unit x) : is_unit (f x) :=
+@[to_additive] lemma map [monoid_hom_class F M N] (f : F) {x : M} (h : is_unit x) : is_unit (f x) :=
 by rcases h with ⟨y, rfl⟩; exact (units.map (f : M →* N) y).is_unit
 
 /-- If a homomorphism `f : M →* N` sends each element to an `is_unit`, then it can be lifted
 to `f : M →* Nˣ`. See also `units.lift_right` for a computable version. -/
 @[to_additive "If a homomorphism `f : M →+ N` sends each element to an `is_add_unit`, then it can be
 lifted to `f : M →+ add_units N`. See also `add_units.lift_right` for a computable version."]
-noncomputable def is_unit.lift_right [monoid M] [monoid N] (f : M →* N)
-  (hf : ∀ x, is_unit (f x)) : M →* Nˣ :=
+noncomputable def lift_right (f : M →* N) (hf : ∀ x, is_unit (f x)) : M →* Nˣ :=
 units.lift_right f (λ x, (hf x).unit) $ λ x, rfl
 
-@[to_additive] lemma is_unit.coe_lift_right [monoid M] [monoid N] (f : M →* N)
-  (hf : ∀ x, is_unit (f x)) (x) :
+@[to_additive] lemma coe_lift_right (f : M →* N) (hf : ∀ x, is_unit (f x)) (x) :
   (is_unit.lift_right f hf x : N) = f x := rfl
 
-@[simp, to_additive] lemma is_unit.mul_lift_right_inv [monoid M] [monoid N] (f : M →* N)
-  (h : ∀ x, is_unit (f x)) (x) : f x * ↑(is_unit.lift_right f h x)⁻¹ = 1 :=
+@[simp, to_additive] lemma mul_lift_right_inv (f : M →* N) (h : ∀ x, is_unit (f x)) (x) :
+  f x * ↑(is_unit.lift_right f h x)⁻¹ = 1 :=
 units.mul_lift_right_inv (λ y, rfl) x
 
-@[simp, to_additive] lemma is_unit.lift_right_inv_mul [monoid M] [monoid N] (f : M →* N)
-  (h : ∀ x, is_unit (f x)) (x) : ↑(is_unit.lift_right f h x)⁻¹ * f x = 1 :=
+@[simp, to_additive] lemma lift_right_inv_mul (f : M →* N) (h : ∀ x, is_unit (f x)) (x) :
+  ↑(is_unit.lift_right f h x)⁻¹ * f x = 1 :=
 units.lift_right_inv_mul (λ y, rfl) x
 
+end monoid
+
+section division_monoid
+variables [division_monoid α] {a b c : α}
+
+@[simp, to_additive] protected lemma inv_mul_cancel : is_unit a → a⁻¹ * a = 1 :=
+by { rintro ⟨u, rfl⟩, rw [←units.coe_inv, units.inv_mul] }
+
+@[simp, to_additive] protected lemma mul_inv_cancel : is_unit a → a * a⁻¹ = 1 :=
+by { rintro ⟨u, rfl⟩, rw [←units.coe_inv, units.mul_inv] }
+
+/-- The element of the group of units, corresponding to an element of a monoid which is a unit. As
+opposed to `is_unit.unit`, the inverse is computable and comes from the inversion on `α`. This is
+useful to transfer properties of inversion in `units α` to `α`. See also `to_units`. -/
+@[to_additive "The element of the additive group of additive units, corresponding to an element of
+an additive monoid which is an additive unit. As opposed to `is_add_unit.add_unit`, the negation is
+computable and comes from the negation on `α`. This is useful to transfer properties of negation in
+`add_units α` to `α`. See also `to_add_units`.", simps]
+def unit' (h : is_unit a) : αˣ := ⟨a, a⁻¹, h.mul_inv_cancel, h.inv_mul_cancel⟩
```

Comment:
My initial impression is that this doesn not belong in a file called `algebra/hom/units`, as it mentions nothing about morphisms.

If there's really nowhere better to put it, can you edit the module docstring to mention that this file contains things not about morphisms too?

Reply:
Yeah I'm fighting the import hierarchy here. I mentioned this in the module docstring and added a todo.